### PR TITLE
fixes #2805

### LINKED ISSUE #2805: Missing function return type for `validatorCompiler`
## 🐛 Bug Report

The current type definition for the validator compiler is missing the function type return type.

```ts
export interface FastifyValidationResult {
  errors?: FastifySchemaValidationError[] | null;
}

export type FastifySchemaCompiler = (routeSchema: FastifyRouteSchemaDef) => FastifyValidationResult
```

It should look more like this to reflect the current JavaScript code:

```ts
export interface FastifyValidatorFunction {
  (data: any): { error?: Error | null, value?: any } | false;
  errors?: FastifySchemaValidationError[] | null;
}

export type FastifySchemaCompiler = (routeSchema: FastifyRouteSchemaDef) => FastifyValidatorFunction
```

Also I'm not sure why the `errors` field exists on the validator function.

## To Reproduce

```js
server.post("/test", {
  validatorCompiler: ({ schema }) => {
    return data => {
      if (!data || data.constructor != Object) {
        return { error: new Error("value is not an object") }
      }
      return { value: data }
    }
  }
}, async req => req.body)
```

Resulting type error:

```
Type '({ schema }: FastifyRouteSchemaDef) => (data: any) => { error: Error; value?: undefined; } | { value: any; error?: undefined; }' is not assignable to type 'FastifySchemaCompiler'.
  Type '(data: any) => { error: Error; value?: undefined; } | { value: any; error?: undefined; }' has no properties in common with type 'FastifyValidationResult'.
```

## Your Environment

- *node version*: 14
- *fastify version*: 3.10.1
- *os*: Linux

### PR DESCRIPTION
**Pull request by VigneshMurugan**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: types/schema.d.ts [hunk 1]
```diff
@@ -26,6 +26,7 @@ export interface FastifySchemaValidationError {
 }
 
 export interface FastifyValidationResult {
+  (data:any): { error?: Error; value?: any } | PromiseLike<any> | boolean;
```
Comment (climba03003) [inline types/schema.d.ts:29]:
```suggestion
  (this: any, data: any): unknown
```

**Review by climba03003 — COMMENTED** [review]
You can take `ajv` [typing](https://github.com/ajv-validator/ajv/blob/ca2ae61c489f45fa2ec3ff2ee78b10136cb1ed3c/lib/types/index.ts#L47) as example.
```
interface ValidateFunction<T = unknown> {
  (this: Ajv | any, data: any, dataCxt?: DataValidationCxt): data is T
  errors?: null | ErrorObject[]
  evaluated?: Evaluated
  schema: AnySchema
  schemaEnv: SchemaEnv
  source?: SourceCode
}
```

File: types/schema.d.ts (hunk 1 above)
Comment (VigneshMurugan) [reply types/schema.d.ts]: Awesome suggestion @climba03003! Thanks

**Review by VigneshMurugan — COMMENTED** [review]
#2805

**Comment by mcollina** [comment]
This needs tests

**Comment by VigneshMurugan** [comment]
> This needs tests

Done

**Comment by VigneshMurugan** [comment]
@mcollina are we waiting for more reviews ?

**Comment by mcollina** [comment]
@VigneshMurugan ideally yes.

cc @fox1t @Ethan-Arrowood

File: types/schema.d.ts [hunk 2]
```diff
@@ -26,6 +26,7 @@ export interface FastifySchemaValidationError {
 }
 
 export interface FastifyValidationResult {
+  (this: any, data: any): unknown;
```
Comment (fox1t) [inline types/schema.d.ts:29]: Why using unknown? I suppose this should be the interface of a `validatorCompiler` that returns a function that returns what is currently implemented as `FastifyValidationResult`. Am I wrong?

Comment (fox1t) [reply types/schema.d.ts:29]: Reading the suggestion at https://github.com/fastify/fastify/pull/2807#pullrequestreview-572409265, the function should return a boolean.

**Review by 3imed-jaberi — APPROVED** [review]

File: types/schema.d.ts (hunk 2 above)
Comment (climba03003) [reply types/schema.d.ts:29]: It should not be returning boolean as fastify accept other returing value such as 
```
{ error: Error, value: any }
```
In here, returning `unknown` provide the highest flexibility for the user to pass in the function.

Comment (climba03003) [reply types/schema.d.ts:29]: You can look at the implementation here.
https://github.com/fastify/fastify/blob/master/lib/validation.js#L70-L76

Comment (climba03003) [reply types/schema.d.ts:29]: Maybe we just change the type back to this one? @VigneshMurugan 
```suggestion
  (data: any): boolean | PromiseLike<any> | { error?: Error, value?: any }
```

Comment (VigneshMurugan) [reply types/schema.d.ts:29]: I initially had it as per your suggestion but was unsure if this is an acceptable type hence went for unknown. If nobody has an objection i can go for your suggestion !

CC: @mcollina @fox1t @Ethan-Arrowood

Comment (fox1t) [reply types/schema.d.ts:29]: Go for it!

**Comment by mcollina** [comment]
ping @Eomm @fox1t

**Review by climba03003 — APPROVED** [review]

**Review by Ethan-Arrowood — APPROVED** [review]

**Pull request merged**